### ISSUE TITLE
The ability to select the default docker network route and configure network route priority

### ISSUE THREAD

**Issue by Daniel-dev22** [docker/roadmap#658]
**Tell us about your request**
A clear and concise description of what you want to happen or the change you would like to see

It would be nice to be able to set the default docker route to a docker network name defined in the compose and define the route priority in either a dictionary or list where the order in the yaml is the order docker would use.

**Which service(s) is this request for?**
Let us know which product(s) you want this for?

Docker compose I think and maybe more under the hood I am not sure.

**Tell us about the problem you're trying to solve. What are you trying to do, and why is it hard?**

There's no way currently to set the default route in docker or to set the priority it uses. Rather the default route and priority is currently selected based on the alphabet of the network name.
That makes for a bad user experience and a lot of confusion when using docker networks.

**Are you currently working around the issue?**
A clear and concise description of any alternative solutions or features you've considered or are using today.

My work around is to create networks but append an alphabet letter `a-example-network` for example so that network is the default network for example and subsequent networks would come after `a`.

**Additional context**

**Comment by YuriyTigiev** [comment]
`There's no way currently to set the default route in docker or to set the priority it uses. Rather the default route and priority is currently selected based on the alphabet of the network name.`

Real ????????????????

**Comment by YuriyTigiev** [comment]
I read this post and it saved me a lot of time. I had spent several days trying to resolve issues with the default routing in Docker between two networks. The problem was that Docker uses routes based on the network name, rather than the order of the networks that I provided. This approach was not logical. I spent a few days investigating why the correct configuration didn't work. It was really surprising to see the approach chosen by the Docker developers.


________________________________

**Comment by mircsicz** [comment]
What a bummer, just wasted a few hours...

Hope this will become configurable in the not so far future!

Even using the describer instead of the name would already be a huge progress, because that way I would not need to rename a centrally defined Network-Name but could just change it in a compose file. like so:

```
    networks:
      a_eth1.50:
        ipv4_address: xx.xx.1y1.94
      b_eth1.90:
        ipv4_address: xx.xx.1zz1.94

networks:
  a_eth1.50:
    external: true
    name: B-name
  b_eth1.90:
    external: true
    name: A-Name
```
best case you'ld add a 'primary_route' parameter that could simply be added to one of the networks...

**Comment by phirestalker** [comment]
If after all this time the devs think we are not defining our container and their networking correctly to need this feature, then PLEASE point us to documentation of how it is supposed to be done.

Hey all, how hard is it to move over to kubernetes with multiple docker files and networks set up? I GUARANTEE that they have a feature like this.

**Comment by tbclark3** [comment]
Being able to set a default route is critical, and for both ipv4 and ipv6. Please implement this feature!

**Comment by webash** [comment]
This has been added [as of v28](https://docs.docker.com/engine/release-notes/28/#2800):

![Image](https://github.com/user-attachments/assets/4f2dceed-9dce-464a-910c-caf2bc1baff0)

However it looks like there is [bug for compose files](https://github.com/docker/compose/issues/12574) that has [been fixed](https://github.com/compose-spec/compose-go/pull/746) but [not released yet](https://docs.docker.com/compose/releases/release-notes/).

**Comment by mircsicz** [comment]
finally...

Thanks a ton!

**Comment by Billos** [comment]
Does it allow us to set an IP as default route ?
For instance having a Wireguard container in the network and using this container as default route.

**Comment by webash** [comment]
> Does it allow us to set an IP as default route ? For instance having a Wireguard container in the network and using this container as default route.

@Billos no. It just specifies the priority of the network interface - highest numerical value priority wins.

**Comment by crankynetman** [comment]
Does anyone on this thread know how to tell when this has actually been released in the docker compose plugin? I'm trying to take advantage of this but it's really unclear to me how to figure out when it actually goes into compose. 

Thanks all!

**Comment by Detavern** [comment]
> Does anyone on this thread know how to tell when this has actually been released in the docker compose plugin? I'm trying to take advantage of this but it's really unclear to me how to figure out when it actually goes into compose.
> 
> Thanks all!

Verified working with Docker 28.1.1 and Compose v2.35.1.

```docker-compose
service:
  dummy:
    networks:
      net-b:
        gw_priority: 10
      net-a: null
```

And it seems that when IPv6 is enabled on any NIC, it takes precedence over IPv4-only NICs for default gateway selection, regardless of the configured `gw_priority` value.

Don't know if it is the expected behavior. It would be better if we could decouple IPv4 and IPv6 gateway selection.

**Comment by divStar** [comment]
@Detavern thank you so much for verifying the availability. Guess I'll see if it's possible to install those (or newer) versions explicitly. I have Alpine Linux running on bare-metal, so I'll have to do some tricks to accomplish that, but it should be possible.

This feature really is crucial as e.g. on Jellyfin I am attaching two networks (internal for UI/QuickConnect, and ipvlan one for DLNA/UPNP capabilities) and I spent half a day to figure out what was going wrong when Jellyfin suddenly couldn't load Plugins, because the ipvlan did not have an Internet connection.

**Comment by Detavern** [comment]
@divStar Glad you like it! Last year, while setting up WireGuard inside a container, I almost facing the same problem.
It worked fine after the setup, still good when connecting to other networks.
But after the container restarting, it failed—turns out the default NIC had switched.
Very impressive.

**Comment by crankynetman** [comment]
Thanks so much @Detavern for confirming!

**Comment by j0nny55555** [comment]
i am noticing the same thing, even when i specify a MAC, one a host restart or a container restart, there are new MACs and alternative IPv6 addresses

the containers are joined to an external macvlan network, and i would prefer to not have to have a macvlan per docker compose stack, i have specific network segments/subnets and i would like to have a macvlan per segment/subnet (per host)

**Comment by sevmonster** [comment]
> And it seems that when IPv6 is enabled on any NIC, it takes precedence over IPv4-only NICs for default gateway selection, regardless of the configured `gw_priority` value.

I ran into this as well. I thought maybe I had a version that was missing https://github.com/compose-spec/compose-go/pull/746, but it's because of the two networks, one had IPv6 enabled while the other didn't. Very annoying and not at all documented as far as I can see.

**Comment by samphonic** [comment]
It seems like it's maybe even a little more muddy than that. I have two networks, `NetA` with *only* IPv6 and `NetB` both v4 and v6. The `gw_priority` flag appears to be ignored and `NetB` is always chosen regardless of the priority value. If I add an IPv4 address to `NetA` then the priority flag is respected (if I add a third network with only one IP, it always loses to the other two). So it seems like maybe whichever interfaces has more addresses wins out, but kind of annoying as that doesn't *seem* to be the intention.

**Comment by sevmonster** [comment]
I believe that's the same as we were experiencing: interfaces must have at least one IPv4 address to be included in the roster for default gateway.

**Comment by thoughtsunificator** [comment]
Docker Compose version 2.40.3

the worst thing about this, what kills me is the alphabetical order thing, thanks @Daniel-dev22  for the tip! It's the second time I have to deal with this, it's on me for not documenting the workaround...

There seem to be documentation about a priority parameter for the cli: https://docs.docker.com/engine/network/#connecting-to-multiple-networks

I could not find any for docker compose: https://docs.docker.com/reference/compose-file/networks/

Given the fact that they close issues where people complain about lack of documentation we cannot expect anything better I guess.

EDIT: Looks like docker compose no longer cares about alphabetical order, it does care about the gw_priority property. I have absolutely no idea how some of my containers have a certain route has default and others have another even though they're similarly ordered.

**Comment by douniwan5788** [comment]
Docker Swarm currently doesn't support `gw_priority` (along with several other Compose parameters), and it also doesn't follow alphabetical order. Are there any known workarounds or planned fixes for this?

**Comment by sproshut** [comment]
Same issue as above. One swarm-scoped `ipvlan` IPv6 network based on configs on all participating nodes (using `--config-from`) takes precedence as the default IPv4 gateway over the second attached overlay network, which should have been used for v4 egress connections. Creating the swarm-scoped network with `--ipv4=false` flag on configuration networks fails with `Error response from daemon: IPv4 cannot be disabled in a Swarm scoped network`, and changing the alphabetical order doesn't seem to work too. Would be nice to have `gw_priority` supported in the Swarm mode too!

---

**UPD**: got it fixed by getting rid of `ipvlan` and making the network dual-stack compatible. Here are the [instructions](https://gist.github.com/sproshut/54fc52fa0699bd88250ac9dcfabd3d3a). May not be a solution to everything, but maybe someone will find this useful.
